addons/l10n_ar/models/account_journal.py

```diff
@@ -164,3 +154,23 @@ def _onchange_set_short_name(self):
         """
         if self.type == 'sale' and self.l10n_ar_afip_pos_number:
             self.code = "%05i" % self.l10n_ar_afip_pos_number
+
+    def write(self, vals):
+        protected_fields = ('type', 'l10n_ar_afip_pos_system', 'l10n_ar_afip_pos_number', 'l10n_latam_use_documents')
+        fields_to_check = [field for field in protected_fields if field in vals]
+
+        if fields_to_check:
+            for journal in self:
+                if (
+                    journal.company_id.account_fiscal_country_id.code != "AR"
+                    or journal.type not in ['sale', 'purchase']
+                    or not self.env['account.move'].search([('journal_id', '=', journal.id), ('posted_before', '=', True)], limit=1)
```

Comment:
Hello @jbw-odoo 

This may be a question of point of view.

For me it's a misuse of the @api.constrains.
They should be used to check consistancies between the fields of the model, not to check if a field can be modified.
One proof it's wrong is that the decorator arguments aren't used in the check.  It shouldn't be the case https://github.com/odoo/odoo/blob/master/odoo/api.py#L108

Your solution is simpler and will certainly be enough.  But those "bad" constrains could cause other unexpected errors in the future.

As you wish ;)